File: stone_lib/src/main/java/com/github/klee0kai/stone/closed/types/TimeHolder.java
package com.github.klee0kai.stone.closed.types;

import com.github.klee0kai.stone.types.IRef;

/**
 * Stone Private class
 */
public class TimeHolder<T> implements IRef<T> {

    private final TimeScheduler timer;
    private T ob = null;
    private ScheduleTask scheduleTask = null;

    public TimeHolder(TimeScheduler timer) {
        this.timer = timer;
    }

    public TimeHolder(TimeScheduler timer, T ob, long holdTime) {
        this.timer = timer;
        hold(ob, holdTime);
    }

    public synchronized T hold(T ob, long holdTime) {
        clearRef();
        this.ob = ob;
        timer.schedule(scheduleTask = new ScheduleTask(holdTime) {
            @Override
            public void run() {
                clearRef();
            }
        });
        return ob;
    }

    public synchronized T get() {
        return ob;
    }

    private synchronized void clearRef() {
        ob = null;
        if (scheduleTask != null) scheduleTask.cancel();
    }

}

File: stone_multiplatform/src/commonMain/kotlin/com/github/klee0kai/stone/__hidden__/types/holders/TimeHolder.kt
package com.github.klee0kai.stone.__hidden__.types.holders

import com.github.klee0kai.stone.weakref.Ref
import com.github.klee0kai.stone.weakref.SoftRef
import kotlinx.coroutines.*

/**
 * Stone Private class
 */
class TimeHolder<T>(
    private var scope: CoroutineScope = CoroutineScope(SupervisorJob()),
    ob: T? = null,
    holdTime: Long? = null,
) : Ref<T?> {

    var ref: Ref<T?>? = null

    init {
        if (ob != null && holdTime != null) {
            hold(ob, holdTime)
        }
    }

    private var holdJob: Job? = null

    fun hold(ob: T?, holdTime: Long): T? {
        clearRef()
        this.ref = SoftRef(ob)
        holdJob?.cancel()
        holdJob = scope.launch {
            delay(holdTime)
            clearRef()
        }
        return ob
    }

    override fun get(): T? = ref?.get()

    private fun clearRef() {
        ref = null
        holdJob?.cancel()
        holdJob = null
    }

}
